fix static cache data type miss-match

jiqing-feng opened the pull request:
Hi @SunMarc . This PR fixed the data type mismatch when using low-precision static cache. The following code can reproduce the bug:
```python
import torch
from transformers import pipeline

model_id = "EleutherAI/gpt-j-6b"
model_kwargs = {"torch_dtype": torch.bfloat16}

pipe = pipeline("text-generation", model=model_id, model_kwargs=model_kwargs)

generation_config = pipe.model.generation_config
generation_config.cache_implementation="static"

print(pipe("I am happy because", generation_config=generation_config))
```

Output:
```
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
Traceback (most recent call last):
  File "/home/jiqingfe/test_gptj.py", line 13, in <module>
    print(pipe("I am happy because", generation_config=generation_config))
  File "/home/jiqingfe/transformers/src/transformers/pipelines/text_generation.py", line 272, in __call__
    return super().__call__(text_inputs, **kwargs)
  File "/home/jiqingfe/transformers/src/transformers/pipelines/base.py", line 1301, in __call__
    return self.run_single(inputs, preprocess_params, forward_params, postprocess_params)
  File "/home/jiqingfe/transformers/src/transformers/pipelines/base.py", line 1308, in run_single
    model_outputs = self.forward(model_inputs, **forward_params)
  File "/home/jiqingfe/transformers/src/transformers/pipelines/base.py", line 1208, in forward
    model_outputs = self._forward(model_inputs, **forward_params)
  File "/home/jiqingfe/transformers/src/transformers/pipelines/text_generation.py", line 370, in _forward
    generated_sequence = self.model.generate(input_ids=input_ids, attention_mask=attention_mask, **generate_kwargs)
  File "/opt/conda/envs/idp/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/jiqingfe/transformers/src/transformers/generation/utils.py", line 2263, in generate
    result = self._beam_search(
  File "/home/jiqingfe/transformers/src/transformers/generation/utils.py", line 3472, in _beam_search
    outputs = self(**model_inputs, return_dict=True)
  File "/opt/conda/envs/idp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/envs/idp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/jiqingfe/transformers/src/transformers/models/gptj/modeling_gptj.py", line 1098, in forward
    transformer_outputs = self.transformer(
  File "/opt/conda/envs/idp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/envs/idp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/jiqingfe/transformers/src/transformers/models/gptj/modeling_gptj.py", line 838, in forward
    outputs = block(
  File "/opt/conda/envs/idp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/envs/idp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/jiqingfe/transformers/src/transformers/models/gptj/modeling_gptj.py", line 453, in forward
    attn_outputs = self.attn(
  File "/opt/conda/envs/idp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/envs/idp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/jiqingfe/transformers/src/transformers/models/gptj/modeling_gptj.py", line 246, in forward
    key, value = layer_past.update(key, value, self.layer_idx, cache_kwargs)
  File "/home/jiqingfe/transformers/src/transformers/cache_utils.py", line 1220, in update
    k_out.index_copy_(2, cache_position, key_states)
RuntimeError: index_copy_(): self and source expected to have the same dtype, but got (self) BFloat16 and (source) Float
```

jiqing-feng commented:
BTW, I suppose transformers missed some static cache tests, do you have any instructions about where can I add this kind of test? Thanks!

SunMarc commented:
All the tests related to the cache are in test_utils.py file. inside GenerationTesterMixin, you will find the test we perform on all models and in `GenerationIntegrationTests`, these are integration tests.

SunMarc commented on src/transformers/models/gptj/modeling_gptj.py:
```diff
@@ -233,8 +233,8 @@ def forward(
 
-        key = key.permute(0, 2, 1, 3)
-        query = query.permute(0, 2, 1, 3)
+        key = key.permute(0, 2, 1, 3).to(value.dtype)
+        query = query.permute(0, 2, 1, 3).to(value.dtype)
```
could you explain why this is needed for this particular model and why this doesn't happen for llama for example ? Many models have approximately the same modeling code.

jiqing-feng replied:
For llama, we can see sin and cos come from position_embeddings (bf16 tensor) which comes from [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L168). You can see llama's rotary embedding converts the data type. But for gptj, the position embeddings come from [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gptj/modeling_gptj.py#L63), it set the data type to float32, so data type miss-match happens when the input data type is bf16 or fp16.

SunMarc reviewed (commented):
Thanks for the bug fix ! Left a comment

jiqing-feng committed:
fix gptj data type missmatch

Signed-off-by: jiqing-feng <jiqing.feng@intel.com>

jiqing-feng committed:
Merge branch 'main' into gptj

jiqing-feng committed:
Merge branch 'main' into gptj

jiqing-feng commented:
Hi @SunMarc . I left the [comment ](https://github.com/huggingface/transformers/pull/34799#discussion_r1849382212)to explain why llama model doesn't have this issue. BTW, I also added the low-precision static cache tests to avoid this kind of issue in the future, please review it. Thanks!

jiqing-feng committed:
add low precision static cache tests

Signed-off-by: jiqing-feng <jiqing.feng@intel.com>

jiqing-feng committed:
fix format

Signed-off-by: jiqing-feng <jiqing.feng@intel.com>

jiqing-feng committed:
fix low-precision static cache tests

jiqing-feng committed:
fix format

Signed-off-by: jiqing-feng <jiqing.feng@intel.com>

jiqing-feng committed:
avoid config change

Signed-off-by: jiqing-feng <jiqing.feng@intel.com>

SunMarc reviewed (approved):
Thanks for the explanation ! LGTM ! Did you run the static cache tests you added to see if there are other models that requires this fix ?

ArthurZucker reviewed (commented):
➕ on marc's comment. The safe way to do this is to cast key and query to the cache's dtype no? And do this in the `cache_utils` rather than at the modeling level!

jiqing-feng commented:
> ```python
> EleutherAI/gpt-j-6b
> ```

Yes, I have applied your suggestions, thanks!



> Thanks for the explanation ! LGTM ! Did you run the static cache tests you added to see if there are other models that requires this fix ?

The CI already contains the tests that I changed, so currently no other models require it. Besides, I have changed it into cache_utils which should be applied for all language models with static cache.

jiqing-feng committed:
Merge branch 'main' into gptj

jiqing-feng committed:
change data type convert in cache copy

Signed-off-by: jiqing-feng <jiqing.feng@intel.com>

SunMarc commented on src/transformers/cache_utils.py:
```diff
@@ -1226,8 +1226,8 @@ def update(
             try:
-                k_out.index_copy_(2, cache_position, key_states)
-                v_out.index_copy_(2, cache_position, value_states)
+                k_out.index_copy_(2, cache_position, key_states.to(k_out.dtype))
+                v_out.index_copy_(2, cache_position, value_states.to(v_out.dtype))
```
You also have to do it in the `NotImplementedError` logic below no ? Maybe do the cast rigth after we computed `k_out` and `v_out`

jiqing-feng replied:
Right, fixed.

jiqing-feng committed:
fix comment

Signed-off-by: jiqing-feng <jiqing.feng@intel.com>

jiqing-feng committed:
cast key value after k v out

Signed-off-by: jiqing-feng <jiqing.feng@intel.com>

jiqing-feng committed:
Merge branch 'main' into gptj

jiqing-feng commented:
Hi @SunMarc , please review the new changes, thanks!

jiqing-feng committed:
Merge branch 'main' into gptj

SunMarc commented:
> Hi @SunMarc , please review the new changes, thanks!

All good from my side. Pinging @ArthurZucker

ArthurZucker reviewed (approved):
Not completely sure we want to test for float32 as it's quite heavy

SunMarc commented:
I think it was testing for float32 initally and @jiqing-feng added coverage for float16 @ArthurZucker

ArthurZucker commented:
Sounds good then, merging!˜
